File: app/src/main/java/com/cheatdatabase/activity/ui/mycheats/MyUnpublishedCheatsFragment.java
package com.cheatdatabase.activity.ui.mycheats;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.cheatdatabase.R;
import com.cheatdatabase.data.model.UnpublishedCheat;
import com.cheatdatabase.listeners.MyUnpublishedCheatsListItemSelectedListener;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * This is the MODEL of MVVM.
 */
public class MyUnpublishedCheatsFragment extends Fragment implements MyUnpublishedCheatsListItemSelectedListener {
    private static final String TAG = "MyUnpublishedCheatsFt";

    private MyUnpublishedCheatsViewModel myUnpublishedCheatsViewModel;
    private MyUnpublishedCheatsListViewAdapter myUnpublishedCheatsListViewAdapter;

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;

    public static MyUnpublishedCheatsFragment newInstance() {
        return new MyUnpublishedCheatsFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.unpublished_cheats_fragment, container, false);
        ButterKnife.bind(this, view);

//         MyUnpublishedCheatsFragment binding =  DataBindingUtil.setContentView(this, R.layout.unpublished_cheats_fragment);


        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        myUnpublishedCheatsViewModel = new ViewModelProvider(this).get(MyUnpublishedCheatsViewModel.class);
        myUnpublishedCheatsViewModel.init();


        setupRecyclerView();

        myUnpublishedCheatsViewModel.getMyUnpublishedCheatsRepository().observe(getActivity(), new Observer<List<UnpublishedCheat>>() {
            @Override
            public void onChanged(List<UnpublishedCheat> unpublishedCheats) {
                Toast.makeText(getContext(), "XXXXX Unpublished Cheats onChanged", Toast.LENGTH_LONG).show();

                myUnpublishedCheatsListViewAdapter.setUnpublishedCheats(unpublishedCheats);
                myUnpublishedCheatsListViewAdapter.notifyDataSetChanged();
            }
        });

    }

    private void setupRecyclerView() {
        if (myUnpublishedCheatsListViewAdapter == null) {
            myUnpublishedCheatsListViewAdapter = new MyUnpublishedCheatsListViewAdapter(this, getActivity());
            RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getActivity());
            recyclerView.setLayoutManager(layoutManager);
            recyclerView.setHasFixedSize(true);
            recyclerView.setAdapter(myUnpublishedCheatsListViewAdapter);
        } else {
            myUnpublishedCheatsListViewAdapter.notifyDataSetChanged();
        }
    }


    @Override
    public void onCheatClicked(UnpublishedCheat cheat) {
        Log.d(TAG, "onCheatClicked: ");
    }

    @Override
    public void onRejectReasonButtonClicked(UnpublishedCheat cheat) {
        Log.d(TAG, "onRejectReasonButtonClicked: ");

    }

    @Override
    public void onDeleteButtonClicked(UnpublishedCheat cheat) {
        Log.d(TAG, "onDeleteButtonClicked: ");
    }
}

File: app/src/main/java/com/cheatdatabase/activity/ui/mycheats/MyUnpublishedCheatsFragment.kt
package com.cheatdatabase.activity.ui.mycheats

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cheatdatabase.R
import com.cheatdatabase.data.model.UnpublishedCheat
import com.cheatdatabase.listeners.MyUnpublishedCheatsListItemSelectedListener
import kotlinx.android.synthetic.main.unpublished_cheats_fragment.view.*

/**
 * This is the MODEL of MVVM.
 */
class MyUnpublishedCheatsFragment : Fragment(), MyUnpublishedCheatsListItemSelectedListener {
    private var myUnpublishedCheatsViewModel: MyUnpublishedCheatsViewModel? = null
    private var myUnpublishedCheatsListViewAdapter: MyUnpublishedCheatsListViewAdapter? = null

    lateinit var recyclerView: RecyclerView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.unpublished_cheats_fragment, container, false)

        recyclerView = view.recycler_view

//        val binding = UnpublishedCheatsFragmentBinding.inflate(layoutInflater)
//        recyclerView = binding.recyclerView

        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        myUnpublishedCheatsViewModel =
            ViewModelProvider(this).get(MyUnpublishedCheatsViewModel::class.java)
        myUnpublishedCheatsViewModel!!.init()

        setupRecyclerView()

        myUnpublishedCheatsViewModel!!.myUnpublishedCheatsRepository.observe(
            activity!!,
            Observer { unpublishedCheats ->
                Toast.makeText(context, "XXXXX Unpublished Cheats onChanged", Toast.LENGTH_LONG)
                    .show()
                myUnpublishedCheatsListViewAdapter!!.setUnpublishedCheats(unpublishedCheats)
                myUnpublishedCheatsListViewAdapter!!.notifyDataSetChanged()
            })
    }

    private fun setupRecyclerView() {
        if (myUnpublishedCheatsListViewAdapter == null) {
            myUnpublishedCheatsListViewAdapter = MyUnpublishedCheatsListViewAdapter(this, activity)
            val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(activity)
            recyclerView.layoutManager = layoutManager
            recyclerView.setHasFixedSize(true)
            recyclerView.adapter = myUnpublishedCheatsListViewAdapter
        } else {
            myUnpublishedCheatsListViewAdapter!!.notifyDataSetChanged()
        }
    }

    override fun onCheatClicked(cheat: UnpublishedCheat) {
        Log.d(TAG, "onCheatClicked: ")
    }

    override fun onRejectReasonButtonClicked(cheat: UnpublishedCheat) {
        Log.d(TAG, "onRejectReasonButtonClicked: ")
    }

    override fun onDeleteButtonClicked(cheat: UnpublishedCheat) {
        Log.d(TAG, "onDeleteButtonClicked: ")
    }

    companion object {
        private const val TAG = "MyUnpublishedCheatsFt"
        fun newInstance(): MyUnpublishedCheatsFragment {
            return MyUnpublishedCheatsFragment()
        }
    }
}